### PR TITLE
neutron: Added option to use L3 HA with Keepalived

### PR DESCRIPTION
**Pull request by dirkmueller**
This is a fixed version of https://github.com/crowbar/crowbar-openstack/pull/1550 which seems to rot without reviews for a while already.

neutron-ha-tool helps moving routers from one node to other active
nodes. This is triggered when the L3 agent on that particular network
node is down. This behavior is not compatible with L3 HA.

L3 HA implementation uses one master router and several others in
standby. If a router/agent is down, it will be discovered through VRRP
protocol [1] and another one will take the master role.

Both L3 HA and neuton-ha-tool are incompatible because they both has
their own way of balancing routers on fallen nodes.

This implementation lets the user configure L3 HA, which will disable
neutron-ha-tool. If L3 HA is deactivated, neutron-ha-tool comes back.

Notice that switching between neutron-ha-tool and L3 HA is possible, but
the routers created are not updated to reflect this change. If they were
created as HA routers they will remain that way. The contrary is also
true. It is up to the user to manually change those routers or delete
them and create new ones.

Co-Authored-By: Itxaka <igarcia@suse.com>
Co-Authored-By: Abel Navarro <abel.navarro@gmail.com>

[1] https://wiki.openstack.org/wiki/Neutron/L3_High_Availability_VRRP

### CODE REVIEW

File: chef/cookbooks/neutron/recipes/common_agent.rb [hunk 1]
```diff
@@ -335,6 +335,8 @@
         periodic_interval: 40,
         periodic_fuzzy_delay: 5,
         dvr_enabled: neutron[:neutron][:use_dvr],
+        l3_ha: node.roles.include?("neutron-network") && neutron[:neutron][:l3_ha][:use_l3_ha],
```
Comment (toabctl) [inline chef/cookbooks/neutron/recipes/common_agent.rb]: nit: use `l3_ha_enabled` (similar to `dvr_enabled` and similar to what you do in common_config.rb)

File: chef/cookbooks/neutron/recipes/network_agents_ha.rb [hunk 2]
```diff
@@ -134,29 +134,58 @@
 end
 
 if use_l3_agent
-  # Add pacemaker resource for neutron-l3-ha-service
-  ha_service_transaction_objects = []
   ha_service_primitive_name = "neutron-l3-ha-service"
+  if node[:neutron][:l3_ha][:use_l3_ha]
+    ## Do we really need to delete it? what about routers not
```
Comment (toabctl) [inline chef/cookbooks/neutron/recipes/network_agents_ha.rb]: Who can answer that question? @rhafer @vuntz ?

File: chef/cookbooks/neutron/recipes/network_agents_ha.rb [hunk 3]
```diff
@@ -134,29 +134,58 @@
...
+    # marked with --ha Enabled ?!
+    pacemaker_primitive ha_service_primitive_name do
+      agent "systemd:neutron-l3-ha-service"
+      op node[:neutron][:ha][:neutron_l3_ha_resource][:op]
+      action [:stop, :delete]
+      only_if { CrowbarPacemakerHelper.is_cluster_founder?(node) }
+      only_if "crm configure show #{ha_service_primitive_name}"
+    end
```
Comment (toabctl) [inline chef/cookbooks/neutron/recipes/network_agents_ha.rb]: Do we need to call `pacemaker_transaction` to really delete it? Or is that working as it is?

File: chef/cookbooks/neutron/recipes/network_agents_ha.rb [hunk 4]
```diff
@@ -134,29 +134,58 @@
...
 
-  pacemaker_transaction "neutron ha service" do
-    cib_objects ha_service_transaction_objects
-    # note that this will also automatically start the resources
-    action :commit_new
-    only_if { CrowbarPacemakerHelper.is_cluster_founder?(node) }
+    rabbit_settings = fetch_rabbitmq_settings
+
+    crowbar_pacemaker_order_only_existing "o-#{ha_service_primitive_name}" do
+      # While neutron-ha-tool technically doesn't directly depend on postgresql or
```
Comment (toabctl) [inline chef/cookbooks/neutron/recipes/network_agents_ha.rb]: postgresql is no longer used.

File: chef/cookbooks/neutron/recipes/network_agents_ha.rb [hunk 5]
```diff
@@ -134,29 +134,58 @@
...
+      # rabbitmq, if these bits are not running, then neutron-server can run but
+      # can't do what it's being asked. Note that neutron-server does have a
+      # constraint on these services, but it's optional, not mandatory (because it
+      # doesn't need to be restarted when postgresql or rabbitmq are restarted).
+      # So explicitly depend on postgresql and rabbitmq (if they are in the cluster).
+      ordering "( postgresql #{rabbit_settings[:pacemaker_resource]} g-haproxy "\
```
Comment (toabctl) [inline chef/cookbooks/neutron/recipes/network_agents_ha.rb]: no more postgresql

File: chef/cookbooks/neutron/templates/default/l3_agent.ini.erb [hunk 6]
```diff
@@ -1,5 +1,6 @@
 [DEFAULT]
 interface_driver = <%= @interface_driver %>
+ha_vrrp_auth_password = <%= @l3_ha_password %>
```
Comment (toabctl) [inline chef/cookbooks/neutron/templates/default/l3_agent.ini.erb]: shouldn't that be `l3_ha_vrrp_password` ?

File: crowbar_framework/app/models/neutron_service.rb [hunk 7]
```diff
@@ -125,6 +125,7 @@ def create_proposal
     base["attributes"]["neutron"]["service_password"] = random_password
     base["attributes"]["neutron"]["memcache_secret_key"] = random_password
     base["attributes"][@bc_name][:db][:password] = random_password
+    base["attributes"][@bc_name][:l3_ha][:password] = random_password
```
Comment (toabctl) [inline crowbar_framework/app/models/neutron_service.rb]: must be `vrrp_password`

File: chef/data_bags/crowbar/migrate/neutron/304_add_use_l3_ha.rb [hunk 8]
```diff
@@ -0,0 +1,21 @@
+def upgrade(ta, td, a, d)
+  unless a.key? "l3_ha"
+    a["l3_ha"] = ta["l3_ha"]
+  end
+
+  unless defined?(@@neutron_l3_ha_password)
+    service = ServiceObject.new "fake-logger"
+    @@neutron_l3_ha_password = service.random_password
+  end
+  a["l3_ha"]["password"] = @@neutron_l3_ha_password unless a["l3_ha"].key?("password")
```
Comment (toabctl) [inline chef/data_bags/crowbar/migrate/neutron/304_add_use_l3_ha.rb]: must use `vrrp_password`

File: chef/data_bags/crowbar/migrate/neutron/304_add_use_l3_ha.rb [hunk 9]
```diff
@@ -0,0 +1,21 @@
+def upgrade(ta, td, a, d)
+  unless a.key? "l3_ha"
+    a["l3_ha"] = ta["l3_ha"]
+  end
```
Comment (toabctl) [inline chef/data_bags/crowbar/migrate/neutron/304_add_use_l3_ha.rb]: Just add the random password creation in this `unless` branch. that should be enough.

File: chef/cookbooks/neutron/recipes/common_agent.rb (hunk 1 above)
Comment (dirkmueller) [reply chef/cookbooks/neutron/recipes/common_agent.rb]: done

File: chef/cookbooks/neutron/recipes/network_agents_ha.rb (hunk 4 above)
Comment (dirkmueller) [reply chef/cookbooks/neutron/recipes/network_agents_ha.rb]: ups

File: chef/cookbooks/neutron/templates/default/l3_agent.ini.erb (hunk 6 above)
Comment (dirkmueller) [reply chef/cookbooks/neutron/templates/default/l3_agent.ini.erb]: yep

File: chef/cookbooks/neutron/recipes/network_agents_ha.rb (hunk 5 above)
Comment (dirkmueller) [reply chef/cookbooks/neutron/recipes/network_agents_ha.rb]: rebase error

File: chef/data_bags/crowbar/migrate/neutron/304_add_use_l3_ha.rb (hunk 8 above)
Comment (dirkmueller) [reply chef/data_bags/crowbar/migrate/neutron/304_add_use_l3_ha.rb]: done

File: chef/data_bags/crowbar/migrate/neutron/304_add_use_l3_ha.rb (hunk 9 above)
Comment (dirkmueller) [reply chef/data_bags/crowbar/migrate/neutron/304_add_use_l3_ha.rb]: done

File: crowbar_framework/app/models/neutron_service.rb (hunk 7 above)
Comment (dirkmueller) [reply crowbar_framework/app/models/neutron_service.rb]: agreed

File: chef/cookbooks/neutron/recipes/network_agents_ha.rb (hunk 5 above)
Comment (dirkmueller) [reply chef/cookbooks/neutron/recipes/network_agents_ha.rb]: unrelated, I've fixed it in a followup commit

File: chef/cookbooks/neutron/recipes/network_agents_ha.rb (hunk 2 above)
Comment (rhafer) [reply chef/cookbooks/neutron/recipes/network_agents_ha.rb]: I think deleting the resource is fine. If a user explicitly creates a router with `--noha` I think it's a fair assumption that there won't be any automatic failover for that router by the old `neutron-ha-tool` based mechanism.

IMO, even if theoretically possible, we shouldn't allow parallel setups of l3ha and neutron-ha-tool.

**Comment by rhafer** [comment]
BTW, should l3ha become the default for master?

**Comment by toabctl** [comment]
>  I assume this has been at least lightly tested for an agent failover?

There seems to be a l3HA test in the neutron-tempest-plugin (`neutron_tempest_plugin.api.admin.test_routers_ha.RoutersTestHA.test_ha_router_creation`) but I don't find this test executed (or skipped) in our setup. would be nice to get this activated.

**Comment by AbelNavarro** [comment]
[L3 HA](https://trello.com/c/OJyN9gjj/155-l3-ha)

**Comment by aspiers** [comment]
[**@toabctl**](https://github.com/toabctl) commented on [February 8, 2019 8:38 AM](https://github.com/crowbar/crowbar-openstack/pull/2005#issuecomment-461729242 "2019-02-08T08:38:15Z - Replied by Github Reply Comments"):
> >  I assume this has been at least lightly tested for an agent failover?
> 
> There seems to be a l3HA test in the neutron-tempest-plugin (I guess not testing the failover) (`neutron_tempest_plugin.api.admin.test_routers_ha.RoutersTestHA.test_ha_router_creation`) but I don't find this test executed (or skipped) in our setup. would be nice to get this activated.

I know I wrote "lightly tested", but whilst I totally agree that it would be nice to get any relevant tempest tests included, I was referring to testing of an agent failover, which [unfortunately right now has to be a manual test](https://storyboard.openstack.org/#!/story/2002129).

**Comment by aspiers** [comment]
[**@rhafer**](https://github.com/rhafer) commented on [February 8, 2019 8:02 AM](https://github.com/crowbar/crowbar-openstack/pull/2005#issuecomment-461721422 "2019-02-08T08:02:48Z - Replied by Github Reply Comments"):
> BTW, should l3ha become the default for master?

Yes it should be the default.  We've carried our own hack for too long already :-)

**Comment by aojea** [comment]
@rktidwell @swami-dvr 

> [**@rhafer**](https://github.com/rhafer) commented on [February 8, 2019 8:02 AM](https://github.com/crowbar/crowbar-openstack/pull/2005#issuecomment-461721422):
> 
> > BTW, should l3ha become the default for master?
> 
> Yes it should be the default. We've carried our own hack for too long already :-)

@rktidwell @swami-dvr can you advise on the maturity of neutron L3 HA with keepalived in Rocky, Pike and Newton before starting the backports?

**Comment by aojea** [comment]
> > I assume this has been at least lightly tested for an agent failover?
> 
> There seems to be a l3HA test in the neutron-tempest-plugin (I guess not testing the failover) (`neutron_tempest_plugin.api.admin.test_routers_ha.RoutersTestHA.test_ha_router_creation`) but I don't find this test executed (or skipped) in our setup. would be nice to get this activated.

Are we going to merge this without any testing?

**Comment by aspiers** [comment]
[**@aojea**](https://github.com/aojea) commented on [February 8, 2019 10:17 AM](https://github.com/crowbar/crowbar-openstack/pull/2005#issuecomment-461755433 "2019-02-08T10:17:14Z - Replied by Github Reply Comments"):
> @rktidwell @swami-dvr can you advise on the maturity of neutron L3 HA with keepalived in Rocky, Pike and Newton before starting the backports?

L3 HA has been around for years.  I co-presented a talk on it at the Tokyo summit in 2015.  Also here is [a spreadsheet I made a long time ago putting the various failure scenarios under the microscope](https://ethercalc.openstack.org/Pike-Neutron-L3-HA).  This was based on in depth talks with upstream at the time, including Asaaf and Armando (the then PTL).  As you can see it has been in pretty good shape since Ocata.  It's not perfect but certainly better than our in-house solution.

**Comment by aspiers** [comment]
[**@aojea**](https://github.com/aojea) commented on [February 8, 2019 10:18 AM](https://github.com/crowbar/crowbar-openstack/pull/2005#issuecomment-461755744 "2019-02-08T10:18:07Z - Replied by Github Reply Comments"):
> Are we going to merge this without any testing?

NO.  ;-)

File: chef/data_bags/crowbar/migrate/neutron/304_add_use_l3_ha.rb [hunk 10]
```diff
@@ -0,0 +1,22 @@
+def upgrade(ta, td, a, d)
+  unless a.key? "l3_ha"
+    a["l3_ha"] = ta["l3_ha"]
+
+    unless defined?(@@neutron_l3_ha_password)
+      service = ServiceObject.new "fake-logger"
+      @@neutron_l3_ha_password = service.random_password
```
Comment (toabctl) [inline chef/data_bags/crowbar/migrate/neutron/304_add_use_l3_ha.rb]: @dirkmueller I think you told me in another PR, that this variable should match the attribute name. So this should be `@@neutron_l3_ha_vrrp_password`

Comment (dirkmueller) [reply chef/data_bags/crowbar/migrate/neutron/304_add_use_l3_ha.rb]: Yes, could you fix it? I am out this week

**Comment by dirkmueller** [comment]
@aspiers luckily co-presenting at the summit is the best way to ensure that the technology is mature. Given the number of critical incidents we get from those customers that use l3 ha I am not sure about that though.

File: chef/data_bags/crowbar/migrate/neutron/304_add_use_l3_ha.rb (hunk 10 above)
Comment (toabctl) [reply chef/data_bags/crowbar/migrate/neutron/304_add_use_l3_ha.rb]: Fixed

**Comment by aspiers** [comment]
[**@dirkmueller**](https://github.com/dirkmueller) commented on [February 11, 2019 9:41 AM](https://github.com/crowbar/crowbar-openstack/pull/2005#issuecomment-462266396 "2019-02-11T09:41:08Z - Replied by Github Reply Comments"):
> @aspiers luckily co-presenting at the summit is the best way to ensure that the technology is mature. Given the number of critical incidents we get from those users that use l3 ha I am not sure about making it default  though.

Moving this to a private email thread.

**Pull request merged**